core/drivers/clk/clk.c

```diff
@@ -202,7 +202,21 @@ void clk_disable(struct clk *clk)
 
 unsigned long clk_get_rate(struct clk *clk)
 {
-	return clk->rate;
+	unsigned long rate = clk->rate;
+
+	if (clk->ops->get_rate) {
+		unsigned long parent_rate =  0UL;
+
+		if (clk->parent)
+			parent_rate = clk_get_rate(clk->parent);
+
+		return clk->ops->get_rate(clk, parent_rate);
+	}
+
+	if (clk->parent)
+		rate = clk_get_rate(clk->parent);
```

Comment:
SO:

unsigned long clk_get_rate(struct clk *clk)
{
	return clk->rate;
	unsigned long rate = clk->rate;

	if (clk->ops->get_rate) {
		unsigned long parent_rate =  0UL;

		if (clk->parent)
			parent_rate = clk->parent->rate;

		return clk->ops->get_rate(clk, parent_rate);
	}

	if (clk->parent)
		rate = clk->parent->rate;

	return rate;
}

Reply:
If `parent_rate = clk->parent->rate;` is enough at line 211, then `clk_get_rate()` should call `clk_compute_rate_no_lock()` instead of re-implementing the same logic.

Reply:
How about this implementation:

```
static void clk_compute_rate_no_lock(struct clk *clk)
{
	unsigned long parent_rate = 0;

	if (clk->parent) {
		clk_compute_rate_no_lock(clk->parent);
		parent_rate = clk->parent->rate;
	}

	if (clk->ops->get_rate)
		clk->rate = clk->ops->get_rate(clk, parent_rate);
	else
		clk->rate = parent_rate;
}

unsigned long clk_get_rate(struct clk *clk)
{
	clk_compute_rate_no_lock(clk);

	return clk->rate;
}
```

Reply:
ok for me

only issue for complicated clock tree  when root clock (grand father clokc) can change
in this case, if the rate of parent of clk->parent  change (clk->parent->rate can be impacted but not updated by clk_compute_rate_no_lock)

Reply:
**With this logic**, `clk_compute_rate_no_lock()` updates the cached rate for **the target clock and**  all **its** ascending clocks, should it should be good, no?

**(edited)**